File: lib/formats/arc/arc_archive.rb
require 'lib/virtual_file'

# ARC archive
# Engine: BGI/Ethornell
# Extension: .arc
# Known games:
# - Higurashi No Naku Koro Ni
module ArcArchive
  module_function

  MAGIC = 'PackFile    '

  def add_cli_help(_arg_parser) end

  def parse_cli_options(_arg_parser, _options) end

  class Unpacker
    def unpack(arc_file, output_files, _options)
      magic = arc_file.read(MAGIC.length)
      fail RecognitionError, 'Not an ARC archive' unless magic == MAGIC

      table = read_table(arc_file)
      read_contents(arc_file, table, output_files)
    end

    def read_table(arc_file)
      num_files = arc_file.read(4).unpack('L')[0]
      fail RecognitionError, 'Bad file count' if num_files * 32 > arc_file.size

      table = []
      num_files.times do
        e = {}

        e[:name] = arc_file.peek(arc_file.tell) { arc_file.read_until_zero }
        arc_file.skip(16)

        e[:origin],
        e[:size] = arc_file.read(8).unpack('LL')
        e[:origin] += + MAGIC.length + 4 + num_files * 32
        arc_file.skip(8)

        if e[:origin] + e[:size] > arc_file.size
          fail RecognitionError, 'Bad offset to file'
        end

        table.push(e)
      end
      table
    end

    def read_contents(arc_file, table, output_files)
      table.each do |e|
        output_files.write do
          data = arc_file.peek(e[:origin]) { arc_file.read(e[:size]) }
          VirtualFile.new(e[:name], data)
        end
      end
    end
  end

  class Packer
    def pack(arc_file, input_files, _options)
      input_files.names.each do |name|
        fail 'File names can contain 16 characters max.' if name.length > 16
      end

      arc_file.write(MAGIC)
      table_pos = arc_file.tell
      write_dummy_table(arc_file, input_files.length)

      table = write_contents(arc_file, input_files)

      arc_file.seek(table_pos)
      write_table(arc_file, table)
    end

    def write_dummy_table(arc_file, file_count)
      arc_file.write("\x00" * 4)
      arc_file.write("\x00" * (file_count * 32))
    end

    def write_table(arc_file, table)
      arc_file.write([table.length].pack('L'))
      table.each do |e|
        arc_file.peek(arc_file.tell) { arc_file.write("\x00" * 16) }
        arc_file.peek(arc_file.tell) { arc_file.write(e[:name]) }
        arc_file.skip(16)
        arc_file.write([e[:origin], e[:size], 0, 0].pack('L4'))
      end
    end

    def write_contents(arc_file, input_files)
      table = []

      current_origin = 0
      input_files.each do |file|
        e = {}
        e[:origin] = current_origin
        e[:size] = file.data.length
        e[:name] = file.name
        arc_file.write(file.data)
        current_origin += e[:size]
        table << e
      end

      table
    end
  end
end

File: src/formats/arc/arc_archive.cc
// ARC archive
//
// Company:   -
// Engine:    BGI/Ethornell
// Extension: .arc
//
// Known games:
// - Higurashi No Naku Koro Ni

#include "formats/arc/arc_archive.h"
#include "formats/gfx/cbg_converter.h"

namespace
{
    const std::string magic("PackFile    ", 12);

    std::unique_ptr<VirtualFile> read_file(
        IO &arc_io,
        const CbgConverter &cbg_converter,
        size_t file_count)
    {
        std::unique_ptr<VirtualFile> file(new VirtualFile);

        size_t old_pos = arc_io.tell();
        file->name = arc_io.read_until_zero();
        arc_io.seek(old_pos + 16);

        size_t offset = arc_io.read_u32_le();
        size_t size = arc_io.read_u32_le();
        offset += magic.size() + 4 + file_count * 32;
        arc_io.skip(8);
        if (offset + size > arc_io.size())
            throw std::runtime_error("Bad offset to file");

        old_pos = arc_io.tell();
        arc_io.seek(offset);
        file->io.write_from_io(arc_io, size);
        arc_io.seek(old_pos);

        cbg_converter.try_decode(*file);

        return file;
    }
}

struct ArcArchive::Internals
{
    CbgConverter cbg_converter;
};

ArcArchive::ArcArchive() : internals(new Internals)
{
}

ArcArchive::~ArcArchive()
{
}

void ArcArchive::add_cli_help(ArgParser &arg_parser) const
{
    internals->cbg_converter.add_cli_help(arg_parser);
}

void ArcArchive::parse_cli_options(ArgParser &arg_parser)
{
    internals->cbg_converter.parse_cli_options(arg_parser);
}

void ArcArchive::unpack_internal(
    VirtualFile &file, OutputFiles &output_files) const
{
    if (file.io.read(magic.size()) != magic)
        throw std::runtime_error("Not an ARC archive");

    size_t file_count = file.io.read_u32_le();
    if (file_count * 32 > file.io.size())
        throw std::runtime_error("Bad file count");

    for (size_t i = 0; i < file_count; i ++)
    {
        output_files.save([&]()
        {
            return read_file(file.io, internals->cbg_converter, file_count);
        });
    }
}
